crates/shared/src/price_estimation/sanitized.rs

```diff
@@ -0,0 +1,311 @@
+                }
+
+                if query.buy_token == BUY_ETH_ADDRESS {
+                    estimations_to_forward.push(Query {
+                        buy_token: self.native_token,
+                        ..*query
+                    });
+                    return EstimationProgress::AwaitingEthEstimation;
+                }
+
+                estimations_to_forward.push(*query);
+                EstimationProgress::AwaitingErc20Estimation
+            })
+            .collect::<Vec<_>>();
+
+        let remaining_estimations = self.inner.estimates(&estimations_to_forward[..]).await;
+        Ok((estimation_progress, remaining_estimations))
+    }
+
+    fn merge_trivial_and_forwarded_estimations(
+        partially_estimated: Vec<EstimationProgress>,
+        forwarded_estimations: Vec<EstimationResult>,
+    ) -> Vec<EstimationResult> {
+        let mut forwarded_estimations = forwarded_estimations.into_iter();
+
+        partially_estimated
+            .into_iter()
+            .map(|progress| match progress {
+                EstimationProgress::TrivialSolution(res) => res,
+                EstimationProgress::AwaitingErc20Estimation => forwarded_estimations
+                    .next()
+                    .expect("there is a result for every forwarded estimation"),
+                EstimationProgress::AwaitingEthEstimation => {
+                    let mut res = forwarded_estimations
+                        .next()
+                        .expect("there is a result for every forwarded estimation")?;
+                    res.gas =
+                        res.gas
+                            .checked_add(GAS_PER_WETH_UNWRAP.into())
+                            .ok_or(anyhow::anyhow!(
+                                "cost of unwrapping ETH would overflow gas price"
+                            ))?;
+                    Ok(res)
+                }
+            })
+            .collect()
+    }
+}
+
+#[async_trait::async_trait]
+impl<T: PriceEstimating> PriceEstimating for SanitizedPriceEstimator<T> {
+    async fn estimates(&self, queries: &[Query]) -> Vec<EstimationResult> {
+        match self.bulk_estimate_prices(queries).await {
+            Err(err) => vec![
+                Err(PriceEstimationError::Other(crate::clone_anyhow_error(&err)));
+                queries.len()
+            ],
+            Ok((trivial_estimations, forwarded_estimations)) => {
+                SanitizedPriceEstimator::<T>::merge_trivial_and_forwarded_estimations(
+                    trivial_estimations,
+                    forwarded_estimations,
+                )
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::bad_token::{MockBadTokenDetecting, TokenQuality};
+    use crate::price_estimation::MockPriceEstimating;
+    use model::order::OrderKind;
+    use primitive_types::{H160, U256};
+
+    const BAD_TOKEN: H160 = H160([0x12; 20]);
+
+    #[tokio::test]
+    async fn works() {
```

Comment:
naming-nit: Maybe the test name could be more descriptive. (Don't have any good suggestions though 😛).

Reply:
Done